code/modules/robotics/bot/medbot.dm

```diff
@@ -474,110 +494,194 @@
+		else if (src.emagged) //Emagged! Time to poison everybody.
+			var/reag_check = pick(src.dangerous_stuff)
+			if(!C.reagents.has_reagent(reag_check, src.dangerous_stuff[reag_check] * 1.5)) // *shrug* two-ish doses of our poison, on average
+				reagent_id = src.dangerous_stuff
 
-	if (!reagent_id) //If they don't need any of that they're probably cured!
-		src.oldpatient = src.patient
-		src.patient = null
-		src.currently_healing = 0
-		src.last_found = world.time
+		else
+			if (length(reagent_id) < 1)
+				if (brute >= heal_threshold)
+					if(!C.reagents.has_reagent(src.treatment_brute))
+						reagent_id[src.treatment_brute] = 15
+				if (burn >= heal_threshold)
+					if(!C.reagents.has_reagent(src.treatment_fire))
+						reagent_id[src.treatment_fire] = 15
+				if (C.get_toxin_damage() >= heal_threshold)
+					if(!C.reagents.has_reagent(src.treatment_tox))
+						reagent_id[src.treatment_tox] = 15
+				if(!C.reagents.has_reagent(src.treatment_virus))
+					reagent_id[src.treatment_virus] = 15
+				if (C.get_oxygen_deprivation() >= (15 + heal_threshold))
+					if(!C.reagents.has_reagent(src.treatment_oxy))
+						reagent_id[src.treatment_oxy] = 15
+
+	if (length(reagent_id) < 1) //If they don't need any of that they're probably cured!
+		if(src.terrifying || src.emagged)
+			ON_COOLDOWN(src, "[MEDBOT_HOSTILE_COOLDOWN]-[ckey(src.patient?.name)]", src.hostile_last_patient_cooldown)
+		else
+			ON_COOLDOWN(src, "[MEDBOT_LASTPATIENT_COOLDOWN]-[ckey(src.patient?.name)]", src.last_patient_cooldown)
 		var/message = pick("All patched up!","An apple a day keeps me away.","Feel better soon!")
 		src.speak(message)
-		return
+		src.KillPathAndGiveUp(1)
+		return FALSE
 	else
-		src.update_icon(stun = 0, heal = 1)
-		src.visible_message("<span class='alert'><B>[src] is trying to inject [src.patient]!</B></span>")
-		SPAWN_DBG(3 SECONDS)
-			if ((get_dist(src, src.patient) <= 1) && (src.on))
-				if ((reagent_id == "internal_beaker") && (src.reagent_glass) && (src.reagent_glass.reagents.total_volume))
-					src.reagent_glass.reagents.trans_to(src.patient,src.injection_amount) //Inject from beaker instead.
-					src.reagent_glass.reagents.reaction(src.patient, 2, src.injection_amount)
-				else
-					src.patient.reagents.add_reagent(reagent_id,src.injection_amount)
-				src.visible_message("<span class='alert'><B>[src] injects [src.patient] with the syringe!</B></span>")
+		actions.start(new/datum/action/bar/icon/medbot_inject(src, reagent_id), src)
+	return TRUE
 
-			src.update_icon()
-			src.currently_healing = 0
+/obj/machinery/bot/medbot/DoWhileMoving()
+	. = ..()
+	if (src.patient && IN_RANGE(src,src.patient,1))
+		return TRUE
 
-			if (src.terrifying)
+/obj/machinery/bot/medbot/KillPathAndGiveUp(var/give_up)
+	. = ..()
+	if(give_up)
+		src.currently_healing = 0
+		src.oldpatient = src.patient
+		src.patient = null
+		src.oldloc = null
+		src.path = null
+
+/datum/action/bar/icon/medbot_inject
+	duration = 30 SECONDS
```

Comment:
This is far too long in practice.

Somewhere between 5 and 10 seconds seems like a better value

For patients with only one or two types of damage, this is *significantly* slower than the old behavior of making multiple injections spaced a couple seconds apart.